pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/client/template.go

```diff
@@ -20,73 +20,252 @@ import (
+			if err := t.tinkclient.Create(ctx, template); err != nil {
+				return fmt.Errorf("failed to create Template in Tinkerbell cluster: %w", err)
+			}
+
+			return nil
+		}
+
+		return fmt.Errorf("failed to get template %s: %v", ProvisionWorkerNodeTemplate, err)
 	}
 
 	return nil
 }
 
-// Create a Tinkerbell Template.
-func (t *Template) Create(ctx context.Context, template *template.WorkflowTemplate) error {
-	resp, err := t.client.CreateTemplate(ctx, template)
+func getTemplate(hardware *tinkv1alpha1.Hardware, osImageURL string) (string, error) {
+	actions := []Action{
+		createWipeDiskAction(),
+		createStreamUbuntuImageAction(hardwareDisk1, osImageURL),
+		createGrowPartitionAction(hardwareDisk1),
+		createNetworkConfigAction(hardware),
+		createCloudInitConfigAction(),
+		decodeCloudInitFile(hardware.Name),
+	}
+
+	task := Task{
+		Name:       "os-installation",
+		WorkerAddr: "{{.device_1}}",
+		Volumes:    []string{"/dev:/dev", "/dev/console:/dev/console", "/lib/firmware:/lib/firmware:ro"},
+		Actions:    actions,
+	}
+
+	template := Template{
+		Name:          "ubuntu",
+		Version:       "0.1",
+		GlobalTimeout: 1800,
+		Tasks:         []Task{task},
+	}
+	// Marshal task to YAML
+	yamlData, err := yaml.Marshal(template)
 	if err != nil {
-		return fmt.Errorf("creating template in Tinkerbell: %w", err)
+		return "", fmt.Errorf("error marshaling the template to YAML: %w", err)
 	}
 
-	template.Id = resp.GetId()
+	return string(yamlData), nil
+}
 
-	return nil
+func createWipeDiskAction() Action {
+	wipeScript := `apk add --no-cache util-linux
+disks="{{ .Hardware.Disks }}"
+disks=${disks:1:-1}
+for disk in $disks; do
+  for partition in $(ls ${disk}* 2>/dev/null); do
+    if [ -b "${partition}" ]; then
+      echo "Wiping ${partition}..."
+      wipefs -af "${partition}"
+    fi
+  done
+done
+echo "All partitions on ${disks} have been wiped."
+`
+	return Action{
+		Name:    "wipe-disk",
+		Image:   "alpine:3.18",
+		Timeout: 600,
+		Command: []string{"/bin/sh", "-c", wipeScript},
+	}
 }
 
-// Delete a Tinkerbell Template.
-func (t *Template) Delete(ctx context.Context, id string) error {
-	req := &template.GetRequest{
-		GetBy: &template.GetRequest_Id{Id: id},
+func createStreamUbuntuImageAction(destDisk, osImageURL string) Action {
+	return Action{
+		Name:    "stream-ubuntu-image",
+		Image:   "quay.io/tinkerbell-actions/image2disk:v1.0.0",
+		Timeout: 600,
+		Environment: map[string]string{
+			"DEST_DISK":  destDisk,
+			"IMG_URL":    osImageURL,
+			"COMPRESSED": "true",
+		},
 	}
-	if _, err := t.client.DeleteTemplate(ctx, req); err != nil {
-		if err.Error() == sqlErrorString || err.Error() == sqlErrorStringAlt {
-			return fmt.Errorf("template %w", ErrNotFound)
-		}
+}
 
-		return fmt.Errorf("deleting template from Tinkerbell: %w", err)
+func createGrowPartitionAction(destDisk string) Action {
+	return Action{
+		Name:    "grow-partition",
+		Image:   "quay.io/tinkerbell/actions/cexec:latest",
```

Comment:
This could also be pinned to c5bde803d9f6c90f1a9d5e06930d856d1481854c, which is the same as "latest" at the moment. Or is there a reason not to pin this image?

Reply:
Why? There is always enhancements that er com into latest and all our tinkerbell actions are using latest,

Reply:
You pinned every single other image in every single other action. Except this one. Surely there had to be a reason. :) When I look at the code, I might now have accidentally broke it by thinking "oh, someone forgot to pin this, let me just fix that". So a comment why this one action is not pinned to a version would be appreciated.

Reply:
I guess you have overlooked something, while the tags are pinned for the all the images, due the the image repo and path are different 😉 

```
quay.io/tinkerbell/actions/cexec:latest
```

```
quay.io/tinkerbell-actions/writefile:v1.0.0
```

Reply:
Anyway you really would like to have this in, I have no problem with that 😉. Will pin it down